File: archive/2021s/solutions/ImmortalityPotion.py
from sys import stdin, stdout

want = [int(x) for x in stdin.readline().split()]
d = sum(want)
chem = ['a', 'b', 'c']
cups = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]


ans = []
def fill(cup, c):
	ans.append("F " + str(cup) + " " + chem[c])

def empty(cup):
	ans.append("E " + str(cup))

def pour(cup1, cup2):
	ans.append("P " + str(cup1) + " " + str(cup2))

def finish(cup):
	ans.append("! " + str(cup))

def bsearch(c1, amt, c2):
	"""Ends with adding amt1 of c1 and 1-amt of c2 to 3."""
	a = []	#binary representation of amt1
	x = amt
	for i in range(32):
		x *= 2
		if x > 1:
			a.append(1)
			x -= 1
		else:
			a.append(0)

	empty(1)
	empty(2)
	fill(2, c1)
	for i in range(31, -1, -1):
		pour(2, 1)
		empty(1)
		fill(1, c1 if a[i] else c2)
		pour(1, 2)
	pour(2, 1)
	pour(1, 3)

if(want[0] >= want[1] and want[0] >= want[2]):
	bsearch(1, want[1]/d*2, 0)
	bsearch(2, want[2]/d*2, 0)
elif(want[1] >= want[0] and want[1] >= want[2]):
	bsearch(0, want[0]/d*2, 1)
	bsearch(2, want[2]/d*2, 1)
else:
	bsearch(0, want[0]/d*2, 2)
	bsearch(1, want[1]/d*2, 2)

finish(3)
stdout.write(str(len(ans)) + "\n")
stdout.write("\n".join(ans))

File: archive/2021s/solutions/ImmortalityPotion.java
import java.util.*;
import java.io.*;

public class ImmortalityPotion {

	static ArrayList<String> ans;
	static char[] chem;

	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		int[] want = new int[3];
		want[0] = sc.nextInt();
		want[1] = sc.nextInt();
		want[2] = sc.nextInt();
		double d = want[0] + want[1] + want[2];
		chem = new char[] { 'a', 'b', 'c' };
		ans = new ArrayList<>();
		if (want[0] >= want[1] && want[0] >= want[2]) {
			bsearch(1, want[1] / d * 2, 0);
			bsearch(2, want[2] / d * 2, 0);
		} else if (want[1] >= want[0] && want[1] >= want[2]) {
			bsearch(0, want[0] / d * 2, 1);
			bsearch(2, want[2] / d * 2, 1);
		} else {
			bsearch(0, want[0] / d * 2, 2);
			bsearch(1, want[1] / d * 2, 2);
		}
		finish(3);
		System.out.println(ans.size());
		for(String s : ans) {
			System.out.println(s);
		}
		sc.close();
	}

	static void fill(int cup, int a) {
		ans.add("F " + cup + " " + chem[a]);
	}

	static void empty(int cup) {
		ans.add("E " + cup);
	}

	static void pour(int cup1, int cup2) {
		ans.add("P " + cup1 + " " + cup2);
	}

	static void finish(int cup) {
		ans.add("! " + cup);
	}

	static void bsearch(int c1, double amt, int c2) {
		int[] a = new int[32];
		double x = amt;
		for (int i = 0; i < 32; i++) {
			x *= 2;
			if (x > 1) {
				a[i] = 1;
				x -= 1;
			} else {
				a[i] = 0;
			}
		}
		empty(1);
		empty(2);
		fill(2, c1);
		for (int i = 31; i > -1; i--) {
			pour(2, 1);
			empty(1);
			if (a[i] == 1) {
				fill(1, c1);
			} else {
				fill(1, c2);
			}
			pour(1, 2);
		}
		pour(2, 1);
		pour(1, 3);
	}

}